File: app/src/main/java/com/example/android/tiltspot/CamActivity.java
package com.example.android.tiltspot;

import android.app.ProgressDialog;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.PointF;
import android.media.MediaScannerConnection;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Environment;
import android.support.v4.content.FileProvider;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

import android.app.Activity;
import android.content.Intent;
import android.hardware.Camera;
import android.hardware.camera2.CameraDevice;
import android.os.Bundle;
import android.util.Log;
import android.view.Display;
import android.view.KeyEvent;
import android.view.SurfaceHolder;
import android.view.SurfaceView;
import android.view.View;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.Toast;

import ir.mahdi.mzip.zip.ZipArchive;


public class CamActivity extends AppCompatActivity implements SurfaceHolder.Callback{

    private Camera mCamera;
    private CameraPreview mCameraPreview;
    private static int camId = Camera.CameraInfo.CAMERA_FACING_FRONT;


//    private float[] mRotmeterData = new float[3];

    /** Called when the activity is first created. */
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);


        setContentView(R.layout.cam_layout);


        mCamera = getCameraInstance();


        mCameraPreview = new CameraPreview(this, mCamera);
        RelativeLayout preview = (RelativeLayout) findViewById(R.id.camera_preview);
        preview.addView(mCameraPreview);


        Button captureButton = (Button) findViewById(R.id.button_capture);
        captureButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

//                Thread t = new Thread(new Runnable() {
//                    public void run() {

                        mCamera.takePicture(new Camera.ShutterCallback() { @Override public void onShutter() { } }, null, mPicture);
//
//                    }
//                });

//                t.start();
//                try {
//                    t.join();
//                } catch (InterruptedException e) {
//                    e.printStackTrace();
//                }

            }
        });

        Button flipCamButton = (Button) findViewById(R.id.flipCam);

        flipCamButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                mCamera.stopPreview();
                mCamera.release();

                if(camId == Camera.CameraInfo.CAMERA_FACING_BACK){
                    camId = Camera.CameraInfo.CAMERA_FACING_FRONT;
                }
                else {
                    camId = Camera.CameraInfo.CAMERA_FACING_BACK;
                }
                mCamera = Camera.open(camId);

                Camera.Parameters parameters = mCamera.getParameters();
                List<String> focusModes = parameters.getSupportedFocusModes();

//                if (focusModes.contains(Camera.Parameters.FOCUS_MODE_CONTINUOUS_PICTURE)) {
//                    parameters.setFlashMode(Camera.Parameters.FOCUS_MODE_CONTINUOUS_PICTURE);
//                }

                List<String> flashModes = parameters.getSupportedFlashModes();

//                if (flashModes.contains(Camera.Parameters.FLASH_MODE_ON)) {
//                    parameters.setFlashMode(Camera.Parameters.FLASH_MODE_ON);
//                }

                if(camId == Camera.CameraInfo.CAMERA_FACING_BACK){
                    parameters.setRotation(90);
                }
                else {
                    parameters.setRotation(270);
                }


                mCamera.setParameters(parameters);

                mCameraPreview = new CameraPreview(CamActivity.this, mCamera);
                RelativeLayout preview = (RelativeLayout) findViewById(R.id.camera_preview);
                preview.addView(mCameraPreview);

            }
        });





        Button BackButton = (Button) findViewById(R.id.button4);

        BackButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                releaseCamera(true);

                Intent intent = new Intent(CamActivity.this, MainActivity.class);
                startActivity(intent);

            }
        });


    }


    /**
     * Helper method to access the camera returns null if it cannot get the
     * camera or does not exist
     *
     * @return
     */
    private Camera getCameraInstance() {
        Camera camera = null;
        try {


            
//            camera = Camera.open();

            Camera.CameraInfo cameraInfo = new Camera.CameraInfo();
            int cameraCount = Camera.getNumberOfCameras();
            for (int camIdx = 0; camIdx < cameraCount; camIdx++) {
                Camera.getCameraInfo(camIdx, cameraInfo);
                if (cameraInfo.facing == Camera.CameraInfo.CAMERA_FACING_FRONT) {
                    try {
                        camera = Camera.open(camIdx);
                    } catch (RuntimeException e) {
                    }
                }
            }


            Camera.Parameters parameters = camera.getParameters();


            if(camId == Camera.CameraInfo.CAMERA_FACING_BACK){
                parameters.setRotation(90);
            }
            else {
                parameters.setRotation(270);
            }


            List<String> focusModes = parameters.getSupportedFocusModes();

            if (focusModes.contains(Camera.Parameters.FOCUS_MODE_CONTINUOUS_PICTURE)) {
                parameters.setFlashMode(Camera.Parameters.FOCUS_MODE_CONTINUOUS_PICTURE);
            }

            List<String> flashModes = parameters.getSupportedFlashModes();

//            if (flashModes.contains(Camera.Parameters.FLASH_MODE_ON)) {
//                parameters.setFlashMode(Camera.Parameters.FLASH_MODE_ON);
//            }

            camera.setParameters(parameters);


        } catch (Exception e) {
            // cannot get camera or does not exist
        }
        return camera;
    }


    private Camera getCameraInstance2() {
        Camera camera = null;
        try {



//            camera = Camera.open();

            Camera.CameraInfo cameraInfo = new Camera.CameraInfo();
            int cameraCount = Camera.getNumberOfCameras();
            for (int camIdx = 0; camIdx < cameraCount; camIdx++) {
                Camera.getCameraInfo(camIdx, cameraInfo);
                if (cameraInfo.facing == Camera.CameraInfo.CAMERA_FACING_FRONT) {
                    try {
                        camera = Camera.open(camIdx);
                    } catch (RuntimeException e) {
                    }
                }
            }


            Camera.Parameters parameters = camera.getParameters();


            if(camId == Camera.CameraInfo.CAMERA_FACING_BACK){
                parameters.setRotation(90);
            }
            else {
                parameters.setRotation(270);
            }


            camera.setParameters(parameters);


            camera.setParameters(parameters);


        } catch (Exception e) {
            // cannot get camera or does not exist
        }
        return camera;
    }


    Camera.PictureCallback mPicture = new Camera.PictureCallback() {
        @Override
        public void onPictureTaken(byte[] data, Camera camera) {
            File pictureFile = getOutputMediaFile();
            if (pictureFile == null) {
                return;
            }
            try {
                FileOutputStream fos = new FileOutputStream(pictureFile);
                fos.write(data);
                fos.close();
            } catch (FileNotFoundException e) {

            } catch (IOException e) {
            }

            mCamera.startPreview();
        }
    };

    private static File getOutputMediaFile() {
        File mediaStorageDir = new File(
                Environment
                        .getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES),
                "PoseCam");

        // Create a media file name
        String timeStamp = new SimpleDateFormat("dd_HHmmss")
                .format(new Date());
        File mediaFile;


        int aziI = (int)  (MyService.azimuth * 90/1.57);
        String azi = Integer.toString(aziI);

        int PitchI =  (int)  (MyService.pitch * 90/1.57);
        String Pitch = Integer.toString(PitchI);

        int RollI = (int)  (MyService.roll * 90/1.57);
        String Roll = Integer.toString(RollI);

        String fileName = "PC_"+azi+"_"+Pitch+"_"+Roll+"_"+Integer.toString(camId);


        mediaFile = new File(mediaStorageDir.getPath() + File.separator
                 + fileName + "_"+timeStamp+ ".png");



        return mediaFile;
    }




    @Override
    public void surfaceCreated(SurfaceHolder surfaceHolder) {
        releaseCamera(true);

        mCamera = getCameraInstance();


        mCameraPreview = new CameraPreview(this, mCamera);
        RelativeLayout preview = (RelativeLayout) findViewById(R.id.camera_preview);
        preview.addView(mCameraPreview);

    }

    @Override
    public void surfaceChanged(SurfaceHolder surfaceHolder, int i, int i1, int i2) {
        releaseCamera(true);

        mCamera = getCameraInstance();


        mCameraPreview = new CameraPreview(this, mCamera);
        RelativeLayout preview = (RelativeLayout) findViewById(R.id.camera_preview);
        preview.addView(mCameraPreview);


    }


    private void releaseCamera(boolean remove) {
        if (mCamera != null) {
            if (remove)
                mCameraPreview.getHolder().removeCallback(mCameraPreview);
            mCamera.release();
            mCamera = null;
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        releaseCamera(true);
    }


    @Override
    public void surfaceDestroyed(SurfaceHolder surfaceHolder) {

    }
}

File: app/src/main/java/com/example/android/tiltspot/CamActivity.kt
package com.example.android.tiltspot

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.media.AudioManager
import android.media.MediaActionSound
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.camera.core.*
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import kotlinx.android.synthetic.main.cam_layout.*
import java.io.File
import java.nio.ByteBuffer
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors


typealias LumaListener = (luma: Double) -> Unit


class CamActivity : AppCompatActivity() {

    private var imageCapture: ImageCapture? = null
    var lF = CameraSelector.DEFAULT_BACK_CAMERA
    var fM: Int = ImageCapture.FLASH_MODE_AUTO




    private lateinit var outputDirectory: File
    private lateinit var cameraExecutor: ExecutorService

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.cam_layout)

        // Request camera permissions
        if (allPermissionsGranted()) {
            startCamera()
        } else {
            ActivityCompat.requestPermissions(
                    this, REQUIRED_PERMISSIONS, REQUEST_CODE_PERMISSIONS)
        }




        // Set up the listener for take photo button
        camera_capture_button.setOnClickListener {
            takePhoto()

        }

        button4.setOnClickListener { backButtonAction() }

        flipCam.setOnClickListener { flipCamera() }
        flashButton.setOnClickListener { flashButtonAction() }

        outputDirectory = getOutputDirectory()

        cameraExecutor = Executors.newSingleThreadExecutor()
    }



    override fun onRequestPermissionsResult(
            requestCode: Int, permissions: Array<String>, grantResults:
            IntArray) {
        if (requestCode == REQUEST_CODE_PERMISSIONS) {
            if (allPermissionsGranted()) {
                startCamera()
            } else {
                Toast.makeText(this,
                        "Permissions not granted by the user.",
                        Toast.LENGTH_SHORT).show()
                finish()
            }
        }
    }


    private fun backButtonAction(){
        val intent = Intent(this, MainActivity::class.java)
        startActivity(intent)

    }


    private fun flashButtonAction(){


        when (fM) {
            ImageCapture.FLASH_MODE_OFF -> {
                fM = ImageCapture.FLASH_MODE_ON
                val msg = "Flash Mode = ON "
                Toast.makeText(baseContext, msg, Toast.LENGTH_SHORT).show()
            }

            ImageCapture.FLASH_MODE_ON -> {
                fM = ImageCapture.FLASH_MODE_AUTO
                val msg = "Flash Mode = Auto "
                Toast.makeText(baseContext, msg, Toast.LENGTH_SHORT).show()
            }
            ImageCapture.FLASH_MODE_AUTO -> {

                fM = ImageCapture.FLASH_MODE_OFF
                val msg = "Flash Mode = OFF "
                Toast.makeText(baseContext, msg, Toast.LENGTH_SHORT).show()
            }
        }



//
//
//
//        if (fM.equals(ImageCapture.FLASH_MODE_AUTO))
//        {
//            fM = ImageCapture.FLASH_MODE_ON
//            val msg = "Flash Mode = ON "
//            Toast.makeText(baseContext, msg, Toast.LENGTH_SHORT).show()
//            Log.d(TAG, msg)
//        }
//        else if (fM.equals(ImageCapture.FLASH_MODE_ON))
//        {
//            fM== ImageCapture.FLASH_MODE_OFF
//            val msg = "Flash Mode = OFF "
//            Toast.makeText(baseContext, msg, Toast.LENGTH_SHORT).show()
//            Log.d(TAG, msg)
//        }
//        else if (fM.equals(ImageCapture.FLASH_MODE_OFF))
//        {
//            fM== ImageCapture.FLASH_MODE_AUTO
//
//            val msg = "Flash Mode = Auto "
//            Toast.makeText(baseContext, msg, Toast.LENGTH_SHORT).show()
//        }


        startCamera()


    }


    private fun flipCamera(){

        if (lF== CameraSelector.DEFAULT_BACK_CAMERA)
        {
            lF = CameraSelector.DEFAULT_FRONT_CAMERA
        }
        else
        {
            lF = CameraSelector.DEFAULT_BACK_CAMERA
        }
        startCamera()

    }



    private fun takePhoto() {
        // Get a stable reference of the modifiable image capture use case
        val imageCapture = imageCapture ?: return

        // Create time-stamped output file to hold the image

        val mediaStorageDir: File = File(
                Environment
                        .getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES),
                "PoseCam")


        val timeStamp = SimpleDateFormat("dd_HHmmss")
                .format(Date())
        val photoFile: File


        val aziI = (MyService.azimuth * 90 / 1.57).toInt()
        val azi = Integer.toString(aziI)

        val PitchI = (MyService.pitch * 90 / 1.57).toInt()
        val Pitch = Integer.toString(PitchI)

        val RollI = (MyService.roll * 90 / 1.57).toInt()
        val Roll = Integer.toString(RollI)

        val isBack = if ((lF== CameraSelector.DEFAULT_BACK_CAMERA)) 1 else 0

        val fileName = "PC_" + azi + "_" + Pitch + "_" + Roll + "_" + Integer.toString(isBack)


        photoFile = File(mediaStorageDir.getPath().toString() + File.separator
                + fileName + "_" + timeStamp + ".png")





        // Create output options object which contains file + metadata
        val outputOptions = ImageCapture.OutputFileOptions.Builder(photoFile).build()

        // Set up image capture listener, which is triggered after photo has
        // been taken
        imageCapture.takePicture(
                outputOptions, ContextCompat.getMainExecutor(this), object : ImageCapture.OnImageSavedCallback {
            override fun onError(exc: ImageCaptureException) {
                Log.e(TAG, "Photo capture failed: ${exc.message}", exc)
            }

            override fun onImageSaved(output: ImageCapture.OutputFileResults) {
                val savedUri = Uri.fromFile(photoFile)
                val msg = "Photo capture succeeded: $fileName"
                Toast.makeText(baseContext, msg, Toast.LENGTH_SHORT).show()
                Log.d(TAG, msg)
            }
        })
    }





    private fun startCamera() {
        val cameraProviderFuture = ProcessCameraProvider.getInstance(this)

        cameraProviderFuture.addListener(Runnable {
            // Used to bind the lifecycle of cameras to the lifecycle owner
            val cameraProvider: ProcessCameraProvider = cameraProviderFuture.get()

            // Preview
            val preview = Preview.Builder()
                    .build()
                    .also {
                        it.setSurfaceProvider(viewFinder.createSurfaceProvider())
                    }

            imageCapture = ImageCapture.Builder().setFlashMode(fM).setCaptureMode(ImageCapture.CAPTURE_MODE_MAXIMIZE_QUALITY)
                    .build()

            val imageAnalyzer = ImageAnalysis.Builder()
                    .build()
                    .also {
                        it.setAnalyzer(cameraExecutor, LuminosityAnalyzer { luma ->
                        })
                    }

            // Select back camera as a default
            val cameraSelector = lF

            try {
                // Unbind use cases before rebinding
                cameraProvider.unbindAll()

                // Bind use cases to camera
                var camera = cameraProvider.bindToLifecycle(
                        this, cameraSelector, preview, imageCapture, imageAnalyzer)

            } catch (exc: Exception) {
                Log.e(TAG, "Use case binding failed", exc)
            }

        }, ContextCompat.getMainExecutor(this))
    }


    private fun allPermissionsGranted() = REQUIRED_PERMISSIONS.all {
        ContextCompat.checkSelfPermission(
                baseContext, it) == PackageManager.PERMISSION_GRANTED
    }

    private fun getOutputDirectory(): File {
        val mediaDir = externalMediaDirs.firstOrNull()?.let {
            File(it, resources.getString(R.string.app_name)).apply { mkdirs() } }
        return if (mediaDir != null && mediaDir.exists())
            mediaDir else filesDir
    }

    override fun onDestroy() {
        super.onDestroy()
        cameraExecutor.shutdown()
    }

    companion object {
        private const val TAG = "CameraXBasic"
        private const val FILENAME_FORMAT = "yyyy-MM-dd-HH-mm-ss-SSS"
        private const val REQUEST_CODE_PERMISSIONS = 10
        private val REQUIRED_PERMISSIONS = arrayOf(Manifest.permission.CAMERA)
    }


    private class LuminosityAnalyzer(private val listener: LumaListener) : ImageAnalysis.Analyzer {

        private fun ByteBuffer.toByteArray(): ByteArray {
            rewind()    // Rewind the buffer to zero
            val data = ByteArray(remaining())
            get(data)   // Copy the buffer into a byte array
            return data // Return the byte array
        }

        override fun analyze(image: ImageProxy) {

            val buffer = image.planes[0].buffer
            val data = buffer.toByteArray()
            val pixels = data.map { it.toInt() and 0xFF }
            val luma = pixels.average()

            listener(luma)

            image.close()
        }
    }



}